Update Chat Interface examples and add more LLM libraries and API providers

abidlabs opened the pull request:
Fixed demos for OpenAI and Langchain. Added more examples for Llama Index, SambaNova, and Hyperbolic AI

Closes: https://github.com/gradio-app/gradio/issues/10003

freddyaboulton commented on demo/llm_hf_transformers/run.py:
```diff
@@ -0,0 +1,47 @@
+    history_transformer_format = list(zip(history[:-1], history[1:])) + [[message, ""]]
+    stop = StopOnTokens()
+
+    messages = "".join(["".join(["\n<human>:"+item[0], "\n<bot>:"+item[1]])
```
This code is now outdated. you don't have to apply the chat template yourself since transformers supports the `role/content` dictionary conversation format. 

I think the streaming can also be simplified but not sure.

freddyaboulton reviewed (commented):
Awesome PR. I think adding llama-cpp-python would be great for those curious/interested in on-device inference. I think Groq would be nice as well.

freddyaboulton commented:
Can we add these as starters in the playground? Trying to think of ways to make these more visible cc @aliabd 

<img width="1321" alt="image" src="https://github.com/user-attachments/assets/7632fb56-fdcc-4850-b61a-a550c5fe10dd">

abidlabs commented:
> Awesome PR. I think adding llama-cpp-python would be great for those curious/interested in on-device inference. I think Groq would be nice as well.

Great suggestions, I'll add `llama-cpp-python`, Groq, and Claude examples

aliabd commented:
@abidlabs I added a chatbots section to the starter demos on the playground, and put 4 there. You can see how in https://github.com/gradio-app/gradio/pull/10025/commits/757ade7396fcf262599c7a99446f6e2bdd8e3931 if you wanna make other changes. 

Unfortunately, a lot of the new demos aren't gonna work in lite, not sure if there's a workaround here @whitphx ?

abidlabs commented:
openai library doesn't work in Lite?

aliabd commented:
Yes, allegedly because of [jiter](https://pypi.org/project/jiter/0.8.0/) which doesn't have a python wheel. But that might not be the only issue. let me take a look cc @whitphx

aliabd commented on guides/05_chatbots/02_chat_interface_examples.md:
```diff
@@ -1,144 +1,57 @@
-In this Guide, we go through several examples of how to use `gr.ChatInterface` with popular LLM libraries.
+In this Guide, we go through several examples of how to use `gr.ChatInterface` with popular LLM libraries and API providers.
```
This guide is missing a title

Also the filename should use - dashes not _ underscores. 

<img width="432" alt="Screenshot 2024-11-27 at 4 10 57 PM" src="https://github.com/user-attachments/assets/26554ff8-6d34-4f0c-a6d5-a23ef8c8df50">

just change the filename to 02_chatinterface-examples.md

aliabd replied:
```suggestion
# Using Popular LLM libraries and APIs

In this Guide, we go through several examples of how to use `gr.ChatInterface` with popular LLM libraries and API providers.
```

aliabd commented on guides/05_chatbots/02_chat_interface_examples.md:
```diff
@@ -1,144 +1,57 @@
 
-Let's start by using `langchain` on top of `openai` to build a general-purpose streaming chatbot application in 19 lines of code. You'll need to have an OpenAI key for this example (keep reading for the free, open-source equivalent!)
+* [Llama Index](#llama-index)
+* [LangChain](#langchain)
+* [OpenAI](#openai)
+* [Hugging Face `transformers`](#hugging-face-transformers)
+* [SambaNova](#sambanova)
+* [Hyperbolic](#hyperbolic)
```
some of these links are broken and not the correct id

aliabd replied:
```suggestion
* [LangChain](#lang-chain)
* [OpenAI](#open-ai)
* [Hugging Face `transformers`](#hugging-face-transformers)
* [SambaNova](#samba-nova)
* [Hyperbolic](#hyperbolic)
```

aliabd commented on guides/05_chatbots/02_chat_interface_examples.md:
````diff
@@ -1,144 +1,57 @@
-```python
-from openai import OpenAI
-import gradio as gr
+**Note**: For quick prototyping, the community-maintained [langchain-gradio repo](https://github.com/AK391/langchain-gradio)  makes it even easier to build chatbots on top of LangChain.
````
Why not use 'Tip' for these notes to make them more prominent? unfortunately there's an open issue with tip and markdown links. will make a fix for that but for now you'll need to use <a> tags.

aliabd replied:
```suggestion
Tip: For quick prototyping, the community-maintained <a href='https://github.com/AK391/langchain-gradio'>langchain-gradio repo</a>  makes it even easier to build chatbots on top of LangChain.
```

abidlabs replied:
That's exactly the reason why I didn't do Tip, but if it supports html tags, then this sounds good 👍

abidlabs commented:
It'd be really nice if we could support in on the Playground -- haven't checked, but I assume lots of peopel trying to generate code with openai in the playground

aliabd commented on guides/05_chatbots/02_chat_interface_examples.md:
```diff
@@ -1,144 +1,57 @@
-        if chunk.choices[0].delta.content is not None:
-              partial_message = partial_message + chunk.choices[0].delta.content
-              yield partial_message
+**Note**: For quick prototyping, the  [openai-gradio library](https://github.com/gradio-app/openai-gradio) makes it even easier to build chatbots on top of OpenAI models.
```
```suggestion
Tip: For quick prototyping, the  <a href='https://github.com/gradio-app/openai-gradio'>openai-gradio library</a> makes it even easier to build chatbots on top of OpenAI models.
```

aliabd commented on guides/05_chatbots/02_chat_interface_examples.md:
````diff
@@ -1,144 +1,57 @@
 
-gr.ChatInterface(predict, type="messages").launch()
-```
+**Note**: For quick prototyping, the  [sambanova-gradio library](https://github.com/gradio-app/sambanova-gradio) makes it even easier to build chatbots on top of OpenAI models.
````
```suggestion
Tip: For quick prototyping, the  <a href='https://github.com/gradio-app/sambanova-gradio'>sambanova-gradio library</a> makes it even easier to build chatbots on top of OpenAI models.
```

aliabd commented on guides/05_chatbots/02_chat_interface_examples.md:
```diff
@@ -1,144 +1,57 @@
+
+$code_llm_hyperbolic
+
+**Note**: For quick prototyping, the  [hyperbolic-gradio library](https://github.com/HyperbolicLabs/hyperbolic-gradio) makes it even easier to build chatbots on top of OpenAI models.
```
```suggestion
Tip: For quick prototyping, the  <a href='https://github.com/HyperbolicLabs/hyperbolic-gradio'>hyperbolic-gradio library</a> makes it even easier to build chatbots on top of OpenAI models.
```

aliabd reviewed (commented):
left some notes on the guide

aliabd commented:
@abidlabs yeah agreed. definitely common

abidlabs commented:
I fixed the `transformers` example (much shorter now) and added an example for Claude. I'll add the other ones later when I get a chance. 

@aliabd @freddyaboulton lmk if you have other feedback or if this is ok to merge!

aliabd commented on guides/05_chatbots/02_chat_interface_examples.md:
```diff
@@ -1,144 +1,69 @@
+* [SambaNova](#samba-nova)
+* [Hyperbolic](#hyperbolic)
+* [Anthropic's Claude](#anthropics-claude)
+* [Groq](#groq)
+* [Llama.cpp](#llamacpp)
 
```
just remember to remove these if you merge without adding them. also `#llamacpp` should be `#llama-cpp`

aliabd reviewed (commented):
I think you forgot to change the guide filename to "02_chatinterface-examples.md"

There are a bunch of issues with the creating chatbots fast guide, including broken links and bad rendering. I missed it when I was out, but it's currently breaking the website build. So I'm gonna make a separate PR to fix those

aliabd commented:
@abidlabs PR is ready: https://github.com/gradio-app/gradio/pull/10060 (it's targeting this one)

whitphx commented:
I think there are some different cases:
* When the demo only uses `transformers`, it might be able to use `transformers.js.py` on Lite
* When the demo uses libraries that can't be installed on Pyodide, we can't run it on Lite
* When the demo uses API client libraries that can be installed on Pyodide, we might be able to patch them to work on Lite.

Will take a look.

whitphx reviewed (commented):
nit

abidlabs commented on demo/llm_claude/requirements.txt:
```diff
@@ -0,0 +1 @@
+openai>=1.0.0
```
nice catch

abidlabs commented:
> I think you forgot to change the guide filename to "02_chatinterface-examples.md"

What do you mean, that _is_ the current file name? Merged in your PR btw thank you! Lmk if you have any other feedback

aliabd reviewed (approved):
I changed the filename in my PR. anyway everything looks good! thanks @abidlabs!

abidlabs commented:
Oh lol thank you @aliabd!

abidlabs commented:
And @whitphx and @freddyaboulton! I'll get this in so that we can fix the website on the next release
